### PR TITLE
Add dictionary support for Undersampler base class

### PR DESCRIPTION
**Pull request by bchen1116**
Part 1 of #2105 

This PR adds support for accepting dictionary input in the undersampler base class. I will add support for adding the dictionary support for Undersampler in a separate PR, since I want to handle the dictionary processing/conversion from ratios to number of samples in the sampler base class, which requires [this doc](https://alteryx.quip.com/CcBCAva03d3a/Sampling-dictionary-for-Oversamplers-and-Undersamplers) to be approved and implemented.

### CODE REVIEW

File: evalml/preprocessing/data_splitters/balanced_classification_sampler.py [hunk 1]
```diff
@@ -68,6 +71,22 @@ def _find_ideal_samples(self, y):
...
+            y (pd.Series): Training data targets
+
+        Returns:
+            (dict): dictionary with undersample target class as key, and number of samples to remove as the value.
+                If we don't need to resample, returns empty dictionary.
+        """
+        y_dict = y.value_counts().to_dict()
+        new_dic = {}
+        for k, v in self.sampling_ratio_dict.items():
+            new_dic[k] = max(y_dict[k] - v, 0)
```
Comment (freddyaboulton) [inline evalml/preprocessing/data_splitters/balanced_classification_sampler.py:87]: @bchen1116 The doc said that we'd accept ratios as the values of the dictionary but it seems like we're still expecting counts. Is that coming in the part 2 pr?

Comment (freddyaboulton) [reply evalml/preprocessing/data_splitters/balanced_classification_sampler.py:87]: Got it. I think I understand the plan now!

**Review by freddyaboulton — APPROVED** [review]
I think this looks great @bchen1116 !

File: evalml/preprocessing/data_splitters/balanced_classification_sampler.py [hunk 2]
```diff
@@ -10,12 +10,14 @@
...
     """Class for balanced classification downsampler."""
 
-    def __init__(self, sampling_ratio=0.25, min_samples=100, min_percentage=0.1, random_seed=0):
+    def __init__(self, sampling_ratio=0.25, sampling_ratio_dict=None, min_samples=100, min_percentage=0.1, random_seed=0):
         """
         Arguments:
             sampling_ratio (float): The smallest minority:majority ratio that is accepted as 'balanced'. For instance, a 1:4 ratio would be
                 represented as 0.25, while a 1:1 ratio is 1.0. Must be between 0 and 1, inclusive. Defaults to 0.25.
 
+            sampling_ratio_dict (dict): A dictionary specifying the desired balanced ratio for each target value. Defaults to None.
```
Comment (angela97lin) [inline evalml/preprocessing/data_splitters/balanced_classification_sampler.py:19]: I think it's important to specify that this overrides `sampling_ratio` if both are set?

**Review by angela97lin — APPROVED** [review]
LGTM! Left nit-picky comment about docstring and how sampling_ratio interacts with sampling_ratio_dict but otherwise 🚢

File: evalml/preprocessing/data_splitters/balanced_classification_sampler.py (hunk 2 above)
Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_sampler.py]: Good call, adding!

**Pull request merged**